summingbird-scalding-test/src/test/scala/com/twitter/summingbird/scalding/BatchedStoreProperties.scala

```diff
@@ -0,0 +1,153 @@
+
+import cascading.flow.{ Flow, FlowDef }
+
+import com.twitter.algebird._
+import com.twitter.algebird.monad._
+import com.twitter.summingbird.batch._
+import com.twitter.summingbird.option.Commutative
+import com.twitter.scalding.{ Source => ScaldingSource, Test => TestMode, _ }
+
+import org.scalacheck._
+import org.scalacheck.Prop._
+import org.scalacheck.Properties
+
+/**
+ * Tests for BatchedStore
+ */
+object BatchedStoreProperties extends Properties("BatchedStore's Properties") {
+
+  implicit def intersectionArb[T: Arbitrary: Ordering]: Arbitrary[Intersection[InclusiveLower, ExclusiveUpper, T]] =
+    Arbitrary {
+      for {
+        l <- Arbitrary.arbitrary[T]
+        u <- Arbitrary.arbitrary[T]
+        if implicitly[Ordering[T]].lt(l, u)
+      } yield Intersection(InclusiveLower(l), ExclusiveUpper(u))
+    }
+
+  implicit val arbTimestamp: Arbitrary[Timestamp] = Arbitrary {
+    Gen.choose(1L, 100000L)
+      .map { Timestamp(_) }
+  }
+
+  def sample[T: Arbitrary]: T = Arbitrary.arbitrary[T].sample.get
+
+  implicit val arbitraryPipeFactory: Arbitrary[PipeFactory[(Int, Int)]] = {
+    Arbitrary {
+      Gen.const {
+        StateWithError[(Interval[Timestamp], Mode), List[FailureReason], FlowToPipe[(Int, Int)]] {
+          (timeMode: (Interval[Timestamp], Mode)) =>
+            {
+              val (time: Interval[Timestamp], mode: Mode) = timeMode
+              val a: FlowToPipe[(Int, Int)] = Reader { (fdM: (FlowDef, Mode)) => TypedPipe.from[(Timestamp, (Int, Int))](Seq((Timestamp(10), (2, 3)))) }
+              Right((timeMode, a))
+            }
+        }
+      }
+    }
+  }
+
+  implicit def timeExtractor[T <: (Long, Any)] = TestUtil.simpleTimeExtractor[T]
+
+  implicit val arbitraryInputWithTimeStampAndBatcher: Arbitrary[(List[(Long, Int)], Batcher, TestStore[Int, Int])] = Arbitrary {
+    Arbitrary.arbitrary[List[Int]]
+      .map { in => in.zipWithIndex.map { case (item: Int, time: Int) => (time.toLong, item) } }
+      .map { in =>
+        val batcher: Batcher = TestUtil.randomBatcher(in)
+        val lastTimeStamp = in.size
+        val testStore = TestStore[Int, Int]("test", batcher, sample[Map[Int, Int]], lastTimeStamp)
+        (in, batcher, testStore)
+      }
+  }
+
+  implicit def arbitraryLocalMode: Arbitrary[Mode] = Arbitrary { Gen.const(Local(true)) }
+
+  property("readAfterLastBatch should return interval starting from the last batch written") = {
+    forAll {
+      (diskPipeFactory: PipeFactory[(Int, Int)],
+      interval: Intersection[InclusiveLower, ExclusiveUpper, Timestamp],
+      inputWithTimeStampAndBatcherAndStore: (List[(Long, Int)], Batcher, TestStore[Int, Int]),
+      mode: Mode) =>
+        val (inputWithTimeStamp, batcher, testStore) = inputWithTimeStampAndBatcherAndStore
+        val result = testStore.readAfterLastBatch(diskPipeFactory)((interval, mode))
+
+        result match {
+          case Right(((readInterval: Intersection[InclusiveLower, ExclusiveUpper, Timestamp], _), _)) => {
```

Comment:
Can we also test the upper bound? What is the property there? Same as interval upper bound? <= that?